### PR TITLE
Pass information to IO streams through IOdirectives

### PR DESCRIPTION
**Pull request by darth-tytus**
Groundwork for the records I/O.

Work in progress.

### CODE REVIEW

File: src/AstTranslator.cpp [hunk 1]
```diff
@@ -104,6 +107,26 @@ void AstTranslator::makeIODirective(IODirectives& ioDirective, const AstRelation
...
+    // Casting due to json11.h type requirements.
+    long long arity{static_cast<long long>(rel->getArity() - auxArityAnalysis->getArity(rel))};
+    long long auxArity{static_cast<long long>(auxArityAnalysis->getArity(rel))};
+
+    Json relJson = Json::object{{"arity", arity}, {"auxArity", auxArity},
+            {"types", Json::array(attributesTypes.begin(), attributesTypes.end())}};
+
+    Json typesystem = Json::object{{name, relJson}};
+
+    ioDirective.set("typesystem", typesystem.dump());
```
Comment (b-scholz) [inline src/AstTranslator.cpp]: Let's change it to "types" rather "typesystem".

File: src/ExplainProvenance.h [hunk 2]
```diff
@@ -185,12 +185,7 @@ class ExplainProvenance {
             return "Relation " + relName + " not found\n";
         }
 
-        // create symbol mask
-        std::vector<RamTypeAttribute> symbolMask(rel->getArity());
-
-        for (size_t i = 0; i < rel->getArity(); i++) {
-            symbolMask.at(i) = RamPrimitiveFromChar(*(rel->getAttrType(i)));
-        }
+        // TODO (darth_tytus): Create a JSON typesystem here?
```
Comment (b-scholz) [inline src/ExplainProvenance.h]: Is this resolved? Could you check that the query/explain system (part of provenance) is still working? (NB: we have tests for those). 

If so, please remove the TODO.

File: src/ExplainProvenanceImpl.h [hunk 3]
```diff
@@ -742,6 +742,8 @@ class ExplainProvenanceImpl : public ExplainProvenance {
             return std::make_tuple(-1, -1, std::vector<RamDomain>());
         }
 
+        // TODO (darth_tytus): update to reflect new types.
+
```
Comment (b-scholz) [inline src/ExplainProvenanceImpl.h]: Is this done?

File: src/ExplainProvenance.h (hunk 2 above)
Comment (darth-tytus) [reply src/ExplainProvenance.h]: It seems this method is not used by anything.
Should I remove it, remove TODO or update it?

File: src/ExplainProvenanceImpl.h (hunk 3 above)
Comment (darth-tytus) [reply src/ExplainProvenanceImpl.h]: This is related to float and unsigned. This fragment should be updated
and also the SouffleInterface should. But it is not strictly related
to this pull request. 
Should I update it now, or leave it with other float/unsigned related
updates (aggregates, user-defined functors...) for the time after the
records I/O?

Comment (darth-tytus) [reply src/ExplainProvenanceImpl.h]: #1274

Comment (b-scholz) [reply src/ExplainProvenanceImpl.h]: Ok - let's revisit when record I/O is implemented. We should not forget.

**Pull request merged**